Content.Server/DeviceLinking/Systems/ObjectSensorSystem.cs

```diff
@@ -0,0 +1,160 @@
+    [Dependency] private readonly SharedPopupSystem _popup = default!;
+    [Dependency] private readonly MobStateSystem _mobState = default!;
+    [Dependency] private readonly DeviceLinkSystem _deviceLink = default!;
+    [Dependency] private readonly UseDelaySystem _useDelay = default!;
+
+    private readonly int ModeCount = Enum.GetValues(typeof(ObjectSensorMode)).Length;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<ObjectSensorComponent, ComponentInit>(OnInit);
+        SubscribeLocalEvent<ObjectSensorComponent, ExaminedEvent>(OnExamined);
+        SubscribeLocalEvent<ObjectSensorComponent, InteractUsingEvent>(OnInteractUsing);
+    }
+
+    public override void Update(float deltaTime)
+    {
+        var query = EntityQueryEnumerator<ObjectSensorComponent>();
+        while (query.MoveNext(out var uid, out var component))
+        {
+            UpdateOutput((uid, component));
+        }
+    }
+
+    private void OnInteractUsing(Entity<ObjectSensorComponent> uid, ref InteractUsingEvent args)
+    {
+        if (args.Handled || !_tool.HasQuality(args.Used, uid.Comp.CycleQuality))
+            return;
+
+        if (TryComp<UseDelayComponent>(uid, out var delay)
+            && !_useDelay.TryResetDelay((uid, delay), true))
+            return;
+
+        var mode = (int) uid.Comp.Mode;
+        mode = ++mode % ModeCount;
+        uid.Comp.Mode = (ObjectSensorMode) mode;
+
+        UpdateOutput(uid);
+
+        _audio.PlayPvs(uid.Comp.CycleSound, uid);
+        var msg = Loc.GetString("object-sensor-cycle", ("mode", uid.Comp.Mode.ToString()));
+        _popup.PopupEntity(msg, uid, args.User);
+    }
+
+    private void OnExamined(Entity<ObjectSensorComponent> uid, ref ExaminedEvent args)
+    {
+        if (!args.IsInDetailsRange)
+            return;
+
+        args.PushMarkup(Loc.GetString("object-sensor-examine", ("mode", uid.Comp.Mode.ToString())));
+    }
+
+    private void OnInit(Entity<ObjectSensorComponent> uid, ref ComponentInit args)
+    {
+        _deviceLink.EnsureSourcePorts(uid, uid.Comp.OutputPort1, uid.Comp.OutputPort2, uid.Comp.OutputPort3, uid.Comp.OutputPort4OrMore);
+    }
+
+    private void SetOut(Entity<ObjectSensorComponent> uid, int port, bool signal)
+    {
+        var component = uid.Comp;
+
+        Logger.Debug($"{uid}, {port}, {signal}");
+
+        switch (port)
+        {
+            case 0:
+                break;
+            case 1:
+                _deviceLink.SendSignal(uid, component.OutputPort1, signal);
+                break;
+            case 2:
+                _deviceLink.SendSignal(uid, component.OutputPort2, signal);
+                break;
+            case 3:
+                _deviceLink.SendSignal(uid, component.OutputPort3, signal);
+                break;
+            default:
+                _deviceLink.SendSignal(uid, component.OutputPort4OrMore, signal);
+                break;
+        }
+    }
+
```

Comment:
This is pretty bad to hardcode cases and fields like this, just use an array, you also have a dummy log you left in the pr.